### PR TITLE
Fix operator precedence in copyStringCritical

### PR DESCRIPTION
**Pull request by mikezhang1234567890**
*data[i] does not resolve to what we want due to operator precedence,
so declare another variable copyArr=*data.

Causes a segfault if compact strings are enabled without this change. 

`j9vm.test.jni.CriticalRegionTest` as part of the `J9vmTest` in functional testing will fail if compact strings are enabled, for example by adding `-XX:+CompactStrings` to `EXTRA_OPTIONS` in a grinder run.

### CODE REVIEW

File: runtime/gc_vlhgc/VLHGCAccessBarrier.cpp [hunk 1]
```diff
@@ -450,7 +450,7 @@ MM_VLHGCAccessBarrier::copyStringCritical(J9VMThread *vmThread, GC_ArrayObjectMo
 	} else {
 		if (isCompressed) {
 			for (jint i = 0; i < length; i++) {
-				*data[i] = (jchar)(U_8)J9JAVAARRAYOFBYTE_LOAD(vmThread, (j9object_t)valueObject, i);
+				(*data)[i] = (jchar)(U_8)J9JAVAARRAYOFBYTE_LOAD(vmThread, (j9object_t)valueObject, i);
```
Comment (pshipton) [inline runtime/gc_vlhgc/VLHGCAccessBarrier.cpp:453]: We should using another variable and take `*data` out of the loop.

**Comment by pshipton** [comment]
Pls provide more information in the description about the test case that fails without this fix.

**Comment by mikezhang1234567890** [comment]
Updated description to mention `j9vm.test.jni.CriticalRegionTest` is the failing test.

File: runtime/gc_vlhgc/VLHGCAccessBarrier.cpp (hunk 1 above)
Comment (mikezhang1234567890) [reply runtime/gc_vlhgc/VLHGCAccessBarrier.cpp]: Added another variable here.

**Comment by pshipton** [comment]
lgtm

**Review by pshipton — APPROVED** [review]

File: runtime/gc_vlhgc/VLHGCAccessBarrier.cpp [hunk 2]
```diff
@@ -449,8 +449,9 @@ MM_VLHGCAccessBarrier::copyStringCritical(J9VMThread *vmThread, GC_ArrayObjectMo
 		functions->setNativeOutOfMemoryError(vmThread, 0, 0);
 	} else {
 		if (isCompressed) {
+			jchar *dataArr = *data;
```
Comment (keithc-ca) [inline runtime/gc_vlhgc/VLHGCAccessBarrier.cpp:452]: This variable could be declared earlier to capture the value returned from `jniArrayAllocateMemoryFromThread()` and used to write to `*data` and replace references to `*data` later in this function.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Review by keithc-ca — CHANGES_REQUESTED** [review]
The commit message mentions `dataArr`, but the code was updated to use `copyArr`: please update the commit message so it is consistent with these changes.

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
jenkins test sanity amac jdk17

**Comment by mikezhang1234567890** [comment]
Looks like one of the testlists in the test run failed to start.

**Comment by pshipton** [comment]
Infra issue. I've done a rebuild on the failed part https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_sanity.functional_aarch64_mac_Personal_testList_0/62/
The other part passed.

**Comment by keithc-ca** [comment]
Retest succeeded.

**Comment by pshipton** [comment]
Pls cherry pick for 0.33

**Pull request merged**